plugins/outputs/iotdb/iotdb.go

```diff
@@ -0,0 +1,287 @@
+				keys = append(keys, field.Key)
+				values = append(values, value)
+				dataTypes = append(dataTypes, datatype)
+			}
+		}
+		// Convert timestamp into specified unit
+		ts, tsError := s.ConvertTimestampOfMetric(metric)
+		if tsError != nil {
+			return nil, tsError
+		}
+		timestampList = append(timestampList, ts)
+		// append all metric data of this record to lists
+		deviceidList = append(deviceidList, metric.Name())
+		measurementsList = append(measurementsList, keys)
+		valuesList = append(valuesList, values)
+		dataTypesList = append(dataTypesList, dataTypes)
+		tagsList = append(tagsList, tags)
+	}
+	var recordsWithTags = &RecordsWithTags{
+		DeviceIDList:     deviceidList,
+		MeasurementsList: measurementsList,
+		ValuesList:       valuesList,
+		DataTypesList:    dataTypesList,
+		TimestampList:    timestampList,
+		TagsList:         tagsList,
+	}
+	return recordsWithTags, nil
+}
+
+// modifiy RecordsWithTags according to 'TreateTagsAs' Configuration
+func (s *IoTDB) ModifiyRecordsWithTags(rwt *RecordsWithTags) error {
+	if s.TreateTagsAs == "Measurements" {
+		// method 1: treate Tag(Key:Value) as measurement
+		for index, tags := range rwt.TagsList { // for each record
+			for _, tag := range tags { // for each tag of this record, append it's Key:Value to measurements
+				datatype, value := s.getDataTypeAndValue(tag.Value)
+				if datatype != client.UNKNOW {
+					rwt.MeasurementsList[index] = append(rwt.MeasurementsList[index], tag.Key)
+					rwt.ValuesList[index] = append(rwt.ValuesList[index], value)
+					rwt.DataTypesList[index] = append(rwt.DataTypesList[index], datatype)
+				}
+			}
+		}
+		return nil
+	} else if s.TreateTagsAs == "DeviceID_subtree" {
+		// method 2: treate Tag(Key:Value) as subtree of device id
+		for index, tags := range rwt.TagsList { // for each record
+			subfix := ""
+			for _, tag := range tags { // for each tag, append it's Value
+				subfix = subfix + "." + tag.Value
+			}
+			rwt.DeviceIDList[index] = rwt.DeviceIDList[index] + subfix
+		}
+		return nil
+	}
+	// something go wrong. This configuration should have been checked in func Init().
+	errorMsg := fmt.Sprintf("IoTDB Configuration Error: unknown TreateTagsAs: %s", s.TreateTagsAs)
+	s.Log.Errorf(errorMsg)
+	return errors.New(errorMsg)
+}
+
+// Write records with tags to IoTDB server
+func (s *IoTDB) WriteRecordsWithTags(rwt *RecordsWithTags) error {
+	// deal with tags
+	modifyError := s.ModifiyRecordsWithTags(rwt)
+	if modifyError != nil {
+		return modifyError
+	}
+	// write to IoTDB server
+	status, err := s.session.InsertRecords(rwt.DeviceIDList, rwt.MeasurementsList,
+		rwt.DataTypesList, rwt.ValuesList, rwt.TimestampList)
+	if status != nil {
+		if verifyResult := client.VerifySuccess(status); verifyResult != nil {
+			s.Log.Info(verifyResult)
+		}
+	}
+	return err
+}
+
+func init() {
+	outputs.Add("iotdb", func() telegraf.Output { return newIoTDB() })
+}
+
+func newIoTDB() *IoTDB {
+	return &IoTDB{}
+}
```

Comment:
Please fuse the two
```suggestion
func init() {
	outputs.Add("iotdb", func() telegraf.Output { return &IoTDB{} })
}
```

Furthermore, your `sample.conf` suggest  e.g. a `timeout` defaulting to `5s`, but I don't see this being set anywhere here... Either change the value in `sample.conf` or set sensible defaults here.

Reply:
You are right. I will set sensible defaults here. Thanks for reminding me.